File: Assignment 26 Feb Twitter live/shash/src/services/bookService.ts
import { Book, BookStats } from '../types/book';
import fs from 'fs/promises';
import path from 'path';
import { SearchQuery, MetadataFilter } from '../types/search';

const dataPath = path.join(__dirname, '../../books.json');
let books: Book[] = [];

async function loadBooks(): Promise<void> {
  try {
    const data = await fs.readFile(dataPath, 'utf-8');
    books = JSON.parse(data);
  } catch (error) {
    console.error('Error loading books:', error);
    books = [];
  }
}

async function saveBooks(): Promise<void> {
  try {
    await fs.writeFile(dataPath, JSON.stringify(books, null, 2));
  } catch (error) {
    console.error('Error saving books:', error);
  }
}

async function getAllBooks(genre?: string): Promise<Book[]> {
  await loadBooks();
  if (!genre) return books;
  return books.filter(book => book.genre.toLowerCase() === genre.toLowerCase());
}

async function getBookById(id: string): Promise<Book | null> {
  await loadBooks();
  return books.find(book => book.id === id) || null;
}

async function addBook(book: Omit<Book, 'id'>): Promise<Book> {
  await loadBooks();
  const newBook: Book = {
    ...book,
    id: String(Number(books[books.length-1].id) + 1) // no duplicate id would be formed
  };
  books.push(newBook);
  await saveBooks();
  return newBook;
}

async function updateRating(id: string, rating: number): Promise<Book | null> {
  await loadBooks();
  const book = books.find(book => book.id === id);
  if (!book) return null;

  book.rating = rating;
  await saveBooks();
  return book;
}

async function getStatistics(): Promise<BookStats> {
  await loadBooks();
  const genres = [...new Set(books.map(book => book.genre))];
  
  const averageRatingByGenre = genres.reduce((acc, genre) => {
    const genreBooks = books.filter(book => book.genre === genre);
    const avgRating = genreBooks.reduce((sum, book) => sum + book.rating, 0) / genreBooks.length;
    acc[genre] = Number(avgRating.toFixed(2));
    return acc;
  }, {} as Record<string, number>);

  const sortedByYear = [...books].sort((a, b) => a.publicationYear - b.publicationYear);

  return {
    averageRatingByGenre,
    oldestBook: sortedByYear[0],
    newestBook: sortedByYear[sortedByYear.length - 1]
  };
}

async function searchBooks(searchQuery: SearchQuery): Promise<Book[]> {
    await loadBooks();
    
    return books.filter(book => {
        const results = searchQuery.filters.map(filter => 
            evaluateFilter(book.metadata, filter)
        );

        return searchQuery.operator === 'AND' 
            ? results.every(result => result)
            : results.some(result => result);
    });
}

function evaluateFilter(metadata: Record<string, number>, filter: MetadataFilter): boolean {
    const value = metadata[filter.field];
    if (value === undefined) return false;

    switch (filter.operator) {
        case '>':
            return value > filter.value;
        case '<':
            return value < filter.value;
        case '==':
            return value === filter.value;
        case '>=':
            return value >= filter.value;
        case '<=':
            return value <= filter.value;
        default:
            return false;
    }
}

export const bookService = {
  getAllBooks,
  getBookById,
  addBook,
  updateRating,
  getStatistics,
  searchBooks,
}; 

File: Assignment 26 Feb Twitter live/AJAYKE/backend/app/services/book_service.py
import json
from app.models.book import Book

class BookService:
    def __init__(self):
        data = []
        with open('books.json', 'r') as file:
            data = json.load(file)
        self.books = data
        self.books = [Book(book['id'], book['title'], book['author'], 
                           book['publicationYear'], book['genre'], book['rating']) 
                      for book in self.books]
        self.next_id = 3

    def get_all_books(self, genre=None, min_rating=0):
        filtered = [book for book in self.books]
        if genre:
            filtered = [book for book in filtered if book.genre.lower() == genre.lower()]
        if min_rating:
            filtered = [book for book in filtered if book.rating >= min_rating]
        return [book.to_dict() for book in filtered]

    def get_book_by_id(self, book_id):
        return next((book.to_dict() for book in self.books if book.id == book_id), None)

    def add_book(self, data):
        book = Book(self.next_id, data['title'], data['author'], 
                   data['publicationYear'], data['genre'], data['rating'])
        self.books.append(book)
        self.next_id += 1
        return book.to_dict()

    def update_rating(self, book_id, rating):
        book = next((book for book in self.books if book.id == book_id), None)
        if book:
            book.rating = rating
            return book.to_dict()
        return None

    def get_statistics(self):
        if not self.books:
            return {}
        
        stats = {
            "total_books": len(self.books),
            "avg_rating_by_genre": {},
            "oldest_book": min(self.books, key=lambda x: x.publication_year).to_dict(),
            "newest_book": max(self.books, key=lambda x: x.publication_year).to_dict()
        }
        
        genres = set(book.genre for book in self.books)
        for genre in genres:
            genre_books = [b.rating for b in self.books if b.genre == genre]
            stats["avg_rating_by_genre"][genre] = sum(genre_books) / len(genre_books)
            
        return stats